File: src/kimi_cli/utils/message.py
from __future__ import annotations

from kosong.message import Message

from kimi_cli.wire.types import AudioURLPart, ImageURLPart, TextPart, VideoURLPart


def message_stringify(message: Message) -> str:
    """Get a string representation of a message."""
    # TODO: this should be merged into `kosong.message.Message.extract_text`
    parts: list[str] = []
    for part in message.content:
        if isinstance(part, TextPart):
            parts.append(part.text)
        elif isinstance(part, ImageURLPart):
            suffix = f":{part.image_url.id}" if part.image_url.id else ""
            parts.append(f"[image{suffix}]")
        elif isinstance(part, AudioURLPart):
            suffix = f":{part.audio_url.id}" if part.audio_url.id else ""
            parts.append(f"[audio{suffix}]")
        elif isinstance(part, VideoURLPart):
            suffix = f":{part.video_url.id}" if part.video_url.id else ""
            parts.append(f"[video{suffix}]")
        else:
            parts.append(f"[{part.type}]")
    return "".join(parts)

File: rust/kagent/src/utils/message.rs
use std::fmt::Write;

use kosong::message::{ContentPart, Message};

pub fn message_stringify(message: &Message) -> String {
    let mut out = String::new();
    for part in &message.content {
        match part {
            ContentPart::Text(part) => out.push_str(&part.text),
            ContentPart::ImageUrl(_) => out.push_str("[image]"),
            ContentPart::AudioUrl(part) => {
                append_media(&mut out, "audio", part.audio_url.id.as_deref())
            }
            ContentPart::VideoUrl(_) => out.push_str("[video]"),
            ContentPart::Think(part) => {
                let _ = write!(out, "[{}]", part.kind);
            }
        }
    }
    out
}

fn append_media(out: &mut String, label: &str, id: Option<&str>) {
    match id {
        Some(value) => {
            let _ = write!(out, "[{}:{}]", label, value);
        }
        None => {
            let _ = write!(out, "[{}]", label);
        }
    }
}
